source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1602 @@
+
+The value of a Nix expression can be inserted into a character string with the dollar-sign and braces (`${ }`).
+
+Example:
+
+```nix
+let
+  name = "Nix";
+in
+"hello ${name}"
+```
+
+    "hello Nix"
+
+Only character strings or values that can be represented as a character string are allowed.
+
+Counter-example:
+
+```nix
+let
+  x = 1;
+in
+"${x} + ${x} = ${x + x}"
+```
+
+    error: cannot coerce an integer to a string
+
+           at «string»:4:2:
+
+                3| in
+                4| "${x} + ${x} = ${x + x}"
+                 |  ^
+                5|
+
+Antiquotation can be arbitrarily nested.
+
+(This can become hard to read, and we recommend to avoid it in practice.)
+
+Example:
+
+```nix
+let
+  a = "no";
+in
+"${a + "${a + " ${a}"}"}"
+```
+
+    "no no no"
+
+
+:::{warning}
+You may encounter strings that use the dollar sign (`$`) before an assigned name, but no braces (`{ }`):
+
+These are *not* antiquotations, but usually denote variables in a shell script.
+
+In such cases, the use of names from the surrounding Nix expression is a coincidence.
+
+Example:
+
+```nix
+let
+  out = "Nix"
+in
+"echo ${out} > $out"
+```
+
+    "echo Nix > $out"
+:::
+
+<!-- TODO: link to escaping rules -->
+
+(functions)=
+## Functions
+
+Functions are everywhere in the Nix language and deserve particular attention.
+
+### Single argument
+
+Functions in Nix language can appear in different forms, but always take exactly one argument.
+Argument and function body are separated by a colon (`:`).
+
+Wherever you see a colon (`:`) in Nix language code:
+- on its left is the function argument
+- on its right is the function body.
+
+Nix functions have no names.
+We say they are anonymous, and call such a function a *lambda*.
+
+Example:
+
+```nix
+x: x + 1
+```
+
+    <LAMBDA>
+
+The `<LAMBDA>` indicates the resulting value is an anonymous function.
+
+We can assign functions a name as to any other value.
+
+Example:
+
+```nix
+let
+  f = x: x + 1
+in f
+```
+
+    <LAMBDA>
+
+### Calling functions
+
+Also known as "function application".
+
+Calling a function with an operand means writing the operand after the function.
+
+Example:
+
+```nix
+(x: x + 1) 1
+```
+
+    2
+
+:::{note}
+Since function and operand are separated by white space, sometimes parantheses (`( )`) are required to distinguish expressions.
+:::
+
+Example:
+
+```nix
+let
+  f = x: x.a;
+in
+f { a = 1; }
+```
+
+    1
+
+The above example calls `f` on a literal attribute set.
+One can also pass operands by name.
+
+Example:
+
+```nix
+let
+  f = x: x.a;
+  v = { a = 1; };
+in
+f v
+```
+
+    1
+
+
+### Chaining arguments
+
+Arguments can be chained by nesting functions.
+
+Such a nested function can be used like a function that takes multiple arguments, but offers additional flexibility.
+
+Example:
+
+```nix
+x: y: x + y
+```
+
+    <LAMBDA>
+
+The above function takes one argument and returns a function `y: x + y` with `x` set to the value of the argument.
```

Comment:
I am unfamiliar with the phrase "chaining arguments". The first thing I do when faced with this is try to research further, but I found it hard to find any references for what is meant by this phrase, whether it is a phrase I should be using in communication with other people, etc. I did web searches to try to figure out if this phrase is specific to Nix, functional programming, etc., and came up short. I did find results that refer to chaining *functions* but not arguments. This suggests to me that perhaps a tutorial should not use this phrase (unless it is used elsewhere in Nix reference material).

If the goal is to introduce Nix using familiar and clear concepts, perhaps title this section "Handling multiple arguments" and then explain in the section content that Nix has no such thing but a similar thing can be accomplished by (insert common functional programming term here).

In particular I don't quite know how to understand `x: y: x + y` at first glance. Beginning by explaining that the expression is equivalent to `x: (y: y + x)` would have helped me. It is obvious in retrospect, but I am still learning! :-) For some reason the parentheses and placing the `x` last also helps me understand. Or, phrasing the sentence "The above function takes one argument `x` and returns a function `y: y + x` with `x` set to the value of the argument.

Reply:
> chaining arguments

Yes, you're right. I kind of made that up. I thought about calling it currying, on the one hand this is already quite technical jargon, and on the other hand it's not even right, as currying is defined as transforming higher-arity functions to nested unary functions.

> "Handling multiple arguments" and then explain in the section content that Nix has no such thing

That's what I'll do.

> Beginning by explaining that the expression is equivalent to x: (y: y + x) would have helped me.

I'll definitely keep arguments in the same order, but I'll add the parenthesized version, it's a good idea.